src/client/pfs.go

```diff
@@ -778,31 +836,100 @@ func (c APIClient) PutFileSplit(repoName string, commitID string, path string, d
 // PutFileURL puts a file using the content found at a URL.
 // The URL is sent to the server which performs the request.
 // recursive allow for recursive scraping of some types URLs for example on s3:// urls.
-func (c APIClient) PutFileURL(repoName string, commitID string, path string, url string, recursive bool, overwrite bool) (retErr error) {
-	putFileClient, err := c.PfsAPIClient.PutFile(c.Ctx())
-	if err != nil {
-		return grpcutil.ScrubGRPC(err)
-	}
-	defer func() {
-		if _, err := putFileClient.CloseAndRecv(); err != nil && retErr == nil {
-			retErr = grpcutil.ScrubGRPC(err)
-		}
-	}()
+func (c *putFileClient) PutFileURL(repoName string, commitID string, path string, url string, recursive bool, overwrite bool) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	var overwriteIndex *pfs.OverwriteIndex
 	if overwrite {
 		overwriteIndex = &pfs.OverwriteIndex{0}
 	}
-	if err := putFileClient.Send(&pfs.PutFileRequest{
+	if err := c.c.Send(&pfs.PutFileRequest{
 		File:           NewFile(repoName, commitID, path),
 		Url:            url,
 		Recursive:      recursive,
 		OverwriteIndex: overwriteIndex,
 	}); err != nil {
 		return grpcutil.ScrubGRPC(err)
 	}
+	if c.oneoff {
```

Comment:
Good call, I was thinking that if the `Send` has already calling `Close` isn't going to make much sense. In many cases it won't because `Close` will just error in the same way, i.e. in the case of network errors. But there are a lot of cases where it will actually do something and in the cases where it errors it won't hurt anything to call `Close`.